fancontrol/portio.cpp

```diff
@@ -19,239 +19,141 @@
+//--------------------------------------------------------------------------
+int FANCONTROL::WaitForFlags(char flags, int onoff, int timeout) {
+	char data;
 
-int verbosity = 0;    // verbosity for the logbuf (0= nothing)
-char lasterrorstring[256] = "",
-logbuf[8192] = "";
-
-
-//-------------------------------------------------------------------------
-// read a byte from the embedded controller (EC) via port io 
-//-------------------------------------------------------------------------
-int FANCONTROL::ReadByteFromEC(int offset, char* pdata) {
-	char data = -1;
-	int numToTrySetup = 50;
-	int iOK = false;
-	int iTimeout = 100;
-	int iTimeoutBuf = 1000;
 	int iTime = 0;
 	int iTick = 10;
 
-	while (numToTrySetup > 0) {
-
-		for (iTime = 0; iTime < iTimeoutBuf; iTime += iTick) {    // wait for full buffers to clear
-			data = (char)ReadPort(EC_CTRLPORT) & 0xff;            // or timeout iTimeoutBuf = 1000
-			if (!(data & (EC_STAT_IBF | EC_STAT_OBF))) break;
-			::Sleep(iTick);
-		}
-
-		WritePort(EC_CTRLPORT, EC_CTRLPORT_READ);            // tell 'em we want to "READ"
-
-		for (iTime = 0; iTime < iTimeout; iTime += iTick) {    // wait for IBF and OBF to clear
-			data = (char)ReadPort(EC_CTRLPORT) & 0xff;
-			if (!(data & (EC_STAT_IBF | EC_STAT_OBF))) {
-				iOK = true;
-				break;
-			}
-
-			::Sleep(iTick);
+	// wait for flags to clear
+	for (iTime = 0; iTime < timeout; iTime += iTick) {
+		data = ReadPort(this->EC_CTRL);
 
-		} // try again after a moment
+		int flagstate = (data & flags) != 0;
+		int	wantedstate = onoff != 0;
 
-		if (!iOK) return 0;
-		iOK = false;
+		if (flagstate == wantedstate)
+			return true;
 
-		WritePort(EC_DATAPORT, offset);                        // tell 'em where we want to read from
-
-		for (iTime = 0; iTime < iTimeout; iTime += iTick) {    // wait for OBF 
-			data = (char)ReadPort(EC_CTRLPORT) & 0xff;
-
-			if ((data & EC_STAT_OBF)) {
-				iOK = true;
-				numToTrySetup = 1;
-				break;
-			}
-			::Sleep(iTick); // check again after a moment
-		}
-
-		// decrement counter. If greater than 0 afterwards,
-		// start the read process over again
-		numToTrySetup -= 1;
+		::Sleep(iTick);
 	}
-
-	if (!iOK) return 0;
-
-	*pdata = ReadPort(EC_DATAPORT);
-
-	if (verbosity > 0) sprintf_s(logbuf + strlen(logbuf), sizeof(logbuf + strlen(logbuf)), "readec: offset= %x, data= %02x\n", offset, *pdata);
-
-	return 1;
+	return false;
 }
 
-
 //-------------------------------------------------------------------------
-// write a byte to the embedded controller (EC) via port io
+// read a byte from the embedded controller (EC) via port io 
 //-------------------------------------------------------------------------
-int FANCONTROL::WriteByteToEC(int offset, char NewData) {
-	char data = -1;
-	char data2 = -1;
-	int iOK = false;
-	int iTimeout = 100;
-	int iTimeoutBuf = 1000;
-	int iTime = 0;
-	int iTick = 10;
-	int numToTrySetup = 5;
+int FANCONTROL::ReadByteFromEC(int offset, char* pdata) {
 
-	while (numToTrySetup > 0) {
```

Comment:
do not remove, this is important as sometimes EC R/W will get blocked by the BIOS and the only workaround is by attempting to access the EC many times

Reply:
But a higher-up function is already reading multiple times and even comparing the results check for equality. But someone had removed that code (which was a very sensibly approach to exactly this problem) in the change from V0.6.2 to V.0.6.3. And later on, someone else then saw the need (obviously) for a retry mechanism. Without the V.0.6.2 code to look at, the V.0.6.3 code was subsequently modified to reintroduce retries, albeit down at the portio level.

I don't care either way, I can revert.

Reply:
Yeah it seems to be working fine without it in my very limited testing so far, I guess we can add it back later if needed